CB-19012 Removed reactor dependency entirely, replaced it with thread…

foldik opened the pull request:
# !!!IMPORTANT!!!

Although the change looks huge (more than 700 files) the actual logic is not a lot.

Most of the change is a simple import change where I replaced the following imports

```java
import reactor.bus.Event;
import reactor.bus.EventBus;
import reactor.rx.Promise;
import reactor.rx.Promises;
import reactor.fn.Consumer;
import reactor.fn.timer.Timer;
import reactor.bus.selector.Selectors;
```
with

```java
import com.sequenceiq.cloudbreak.eventbus.Event;
import com.sequenceiq.cloudbreak.eventbus.EventBus;
import com.sequenceiq.cloudbreak.eventbus.Promise;
import java.util.function.Consumer;
import java.util.Timer;
```
The motivation was to replace reactor's implementation of these classes with simpler implementations so that we don't depend on reactor 2 (which is the cause why we can't upgrade azure sdk version).

The real logic that needs to be reviewed is in the following classes:

__com.sequenceiq.cloudbreak.eventbus.EventBus__

It dispatches events to the specific event handler in an async manner.

__com.sequenceiq.cloudbreak.eventbus.Promise__

It gives a way that one thread can wait for another thread to return an object (be it a result or an exception).

__com.sequenceiq.cloudbreak.eventbus.Event__

Contains event key, event headers and the payload.

horadla23 commented:
retest this please

bergerdenes reviewed (approved):
LGTM

lacikaaa commented on common/src/main/java/com/sequenceiq/cloudbreak/eventbus/Event.java:
```diff
@@ -0,0 +1,81 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+public class Event<T> {
```
could you add toString impl? as I see the original class have it

foldik replied:
Sure.

lacikaaa commented on common/src/main/java/com/sequenceiq/cloudbreak/eventbus/EventBus.java:
```diff
@@ -0,0 +1,110 @@
+        return new Builder();
+    }
+
+    private static class EventRouter {
```
why is this an inner class?

foldik replied:
It's an implementation detail of the EventBus that's why I didn't make it a public class.

lacikaaa replied:
it would make the eventbus implementation cleaner and more readable if it was in a separate class. especially there is already a builder in the same.
I think both of them (bus and router) are missing unit tests

lacikaaa commented on common/src/main/java/com/sequenceiq/cloudbreak/eventbus/Promise.java:
```diff
@@ -0,0 +1,65 @@
+        try {
+            return completableFuture.get(timeout, timeUnit);
+        } catch (TimeoutException e) {
+            throw new RuntimeException(e);
```
we usually use some custom exceptions instead of the builtins. would it make sense to use some exact exceptions instead of pure runtimeex?

foldik replied:
I changed it to throw service specific exception.

lacikaaa commented on flow/src/main/java/com/sequenceiq/flow/core/Flow2Handler.java:
```diff
@@ -355,10 +355,14 @@ private FlowAcceptResult createNewFlow(String key, Payload payload, FlowParamete
             logFlowId(flowId);
+            FlowAcceptResult flowAcceptResult = getFlowAcceptResult(flowChainId, flowParameters.getFlowId());
+            if (isAcceptablePayload(payload)) {
+                ((Acceptable) payload).accepted().accept(flowAcceptResult);
```
shall we log here? similarly on line 411-413

lacikaaa commented on flow/src/main/java/com/sequenceiq/flow/reactor/LoggingErrorHandler.java:
```diff
@@ -1,34 +0,0 @@
- * The main functionality of the subsystem is to provide meaningful error messages to the calling systems.
- */
-@Component
-public class LoggingErrorHandler implements Consumer<Throwable> {
```
is this replaced? or it's not necessary anymore? I don't see any info about it in the commit msg

foldik replied:
This is not needed anymore. I marked it in the commit message as well.

lacikaaa commented on flow/src/main/java/com/sequenceiq/flow/reactor/config/EventBusConfig.java:
```diff
@@ -160,8 +89,53 @@ public MDCCleanerThreadPoolExecutor getPoolExecutor() {
+                LOGGER.info("Dispatcher is not alive there is no need to handle flow failure.", exception);
+            }
+        } catch (Exception e) {
+            LOGGER.error("can't handle flow fail", e);
```
maybe log the event, it could be useful during investigation

lacikaaa replied:
I would say, logging the event could be helpful in the beginning of the method

lacikaaa commented on flow/src/main/java/com/sequenceiq/flow/reactor/config/EventBusConfig.java:
```diff
@@ -160,8 +89,53 @@ public MDCCleanerThreadPoolExecutor getPoolExecutor() {
+        }
+    }
+
+    private void handleUnhandledEvent(Event<?> event, ThreadPoolExecutor threadPoolExecutor) {
```
same here, logging the event could be helpful

lnardai reviewed (approved):
If existing comments are resolved, I'm on board with it! Nice work!

foldik commented on common/src/main/java/com/sequenceiq/cloudbreak/eventbus/EventBus.java:
```diff
@@ -0,0 +1,110 @@
+            }
+        }
+
+        @Subscribe
```
Yes you are right. I removed it.

foldik commented:
@lacikaaa 

I changed the commit message and included more details about the change.

I tested the below scenarios:

- Created environment, datalake, datahub.
- Stopped/started/repaired datalake and datahub.
- Upscaled, downscaled, deleted nodes from datahub.
- Upgraded datahub.

lacikaaa reviewed (approved):
please update the readme in a separate pr. thanks, good job!

lnardai commented:
https://media.giphy.com/media/OCu7zWojqFA1W/giphy.gif

foldik commented:
From the CM logs the failed e2e is because of license issue. Unrelated to my change.

```
2022-11-18 13:18:17,740 ERROR main:com.cloudera.cmf.crypto.LicenseLoaderHelper: Failed to complete verification of License key.
java.lang.NullPointerException
	at com.cloudera.cmf.crypto.ClearSignedFileProcessor.getFileSignature(ClearSignedFileProcessor.java:197)
	at com.cloudera.cmf.crypto.ClearSignedFileProcessor.verifyFile(ClearSignedFileProcessor.java:231)
	at com.cloudera.cmf.crypto.LicenseLoaderHelper.readVanillaLicense(LicenseLoaderHelper.java:169)
	at com.cloudera.cmf.crypto.LicenseLoaderHelper.readLicense(LicenseLoaderHelper.java:246)
	at com.cloudera.server.cmf.bootstrap.EntityManagerFactoryBean.readLicenseFromDiskOrDb(EntityManagerFactoryBean.java:414)
	at com.cloudera.server.cmf.bootstrap.EntityManagerFactoryBean.checkVersionDoFail(EntityManagerFactoryBean.java:387)
	at com.cloudera.server.cmf.bootstrap.EntityManagerFactoryBean.getObject(EntityManagerFactoryBean.java:194)
	at com.cloudera.server.cmf.bootstrap.EntityManagerFactoryBean.getObject(EntityManagerFactoryBean.java:87)
	at org.springframework.beans.factory.support.FactoryBeanRegistrySupport.doGetObjectFromFactoryBean(FactoryBeanRegistrySupport.java:169)
	at org.springframework.beans.factory.support.FactoryBeanRegistrySupport.getObjectFromFactoryBean(FactoryBeanRegistrySupport.java:101)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getObjectForBeanInstance(AbstractBeanFactory.java:1828)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.getObjectForBeanInstance(AbstractAutowireCapableBeanFactory.java:1270)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:261)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
	at org.springframework.beans.factory.support.BeanDefinitionValueResolver.resolveReference(BeanDefinitionValueResolver.java:330)
	at org.springframework.beans.factory.support.BeanDefinitionValueResolver.resolveValueIfNecessary(BeanDefinitionValueResolver.java:113)
	at org.springframework.beans.factory.support.ConstructorResolver.resolveConstructorArguments(ConstructorResolver.java:692)
	at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:197)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1361)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1208)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:556)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:516)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:324)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:322)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:897)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:879)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:551)
	at com.cloudera.server.cmf.Main.bootstrapSpringContext(Main.java:563)
	at com.cloudera.server.cmf.Main.<init>(Main.java:286)
	at com.cloudera.server.cmf.Main.main(Main.java:258)
```
